File: packages/schema/src/event.ts
/**
 * Event-related schemas
 * Matches backend: dto/event.go
 */
import { z } from "zod";
import { sessionIdSchema, timestampSchema } from "./common";

/**
 * Event DTO schema
 * Matches: dto.EventDTO
 */
export const eventSchema = z.object({
  id: z.string(),
  type: z.string(),
  session_id: sessionIdSchema,
  data: z.instanceof(Uint8Array).optional(),
  timestamp: timestampSchema,
});

export type Event = z.infer<typeof eventSchema>;

/**
 * Query events request schema
 * Matches: dto.QueryEventsRequest
 */
export const queryEventsRequestSchema = z
  .object({
    session_id: sessionIdSchema.optional(),
    event_type: z.string().optional(),
    since: timestampSchema.optional(),
    until: timestampSchema.optional(),
    limit: z.number().int().min(1).max(1000).default(100),
    offset: z.number().int().min(0).default(0),
  })
  .refine(
    (data) => {
      // Validate limit
      if (data.limit && (data.limit < 1 || data.limit > 1000)) {
        return false;
      }
      // Validate offset
      if (data.offset && data.offset < 0) {
        return false;
      }
      return true;
    },
    {
      message: "Invalid limit or offset values",
    },
  );

export type QueryEventsRequest = z.infer<typeof queryEventsRequestSchema>;

/**
 * Query events response schema
 * Matches: dto.QueryEventsResponse
 */
export const queryEventsResponseSchema = z.object({
  events: z.array(eventSchema),
  total: z.number().int().min(0),
  limit: z.number().int(),
  offset: z.number().int(),
});

export type QueryEventsResponse = z.infer<typeof queryEventsResponseSchema>;

/**
 * Replay events request schema
 * Matches: dto.ReplayEventsRequest
 */
export const replayEventsRequestSchema = z
  .object({
    session_id: sessionIdSchema.optional(),
    event_type: z.string().optional(),
    since: timestampSchema.optional(),
    until: timestampSchema.optional(),
    dry_run: z.boolean().optional(),
  })
  .refine(
    (data) => {
      // At least one filter must be specified
      return !!(data.session_id || data.event_type || data.since || data.until);
    },
    {
      message:
        "At least one filter (session_id, event_type, since, or until) must be specified",
    },
  );

export type ReplayEventsRequest = z.infer<typeof replayEventsRequestSchema>;

/**
 * Replay events response schema
 * Matches: dto.ReplayEventsResponse
 */
export const replayEventsResponseSchema = z.object({
  success: z.boolean(),
  events_found: z.number().int().min(0),
  events_replayed: z.number().int().min(0),
  events_failed: z.number().int().min(0).optional(),
  dry_run: z.boolean(),
  message: z.string(),
  last_error: z.any().optional(),
});

export type ReplayEventsResponse = z.infer<typeof replayEventsResponseSchema>;

File: apps/server/internal/application/dto/event.go
package dto

import (
	"fmt"
)

// EventDTO represents an event in API responses
type EventDTO struct {
	ID        string `json:"id"`
	Type      string `json:"type"`
	SessionID string `json:"session_id"`
	Data      []byte `json:"data,omitempty"`
	Timestamp string `json:"timestamp"`
}

// QueryEventsRequest represents a request to query events
type QueryEventsRequest struct {
	SessionID string `json:"session_id,omitempty" form:"session_id"`
	EventType string `json:"event_type,omitempty" form:"event_type"`
	Since     string `json:"since,omitempty" form:"since"` // RFC3339 timestamp
	Until     string `json:"until,omitempty" form:"until"` // RFC3339 timestamp
	Limit     int    `json:"limit,omitempty" form:"limit"`
	Offset    int    `json:"offset,omitempty" form:"offset"`
}

// Validate validates the query events request
func (r *QueryEventsRequest) Validate() error {
	// Set default limit if not specified
	if r.Limit == 0 {
		r.Limit = 100
	}

	// Validate limit
	if r.Limit < 0 || r.Limit > 1000 {
		return fmt.Errorf("limit must be between 1 and 1000")
	}

	// Validate offset
	if r.Offset < 0 {
		return fmt.Errorf("offset must be non-negative")
	}

	return nil
}

// QueryEventsResponse represents a response to query events
type QueryEventsResponse struct {
	Events []EventDTO `json:"events"`
	Total  int64      `json:"total"`
	Limit  int        `json:"limit"`
	Offset int        `json:"offset"`
}

// ReplayEventsRequest represents a request to replay events
type ReplayEventsRequest struct {
	SessionID string `json:"session_id,omitempty"`
	EventType string `json:"event_type,omitempty"`
	Since     string `json:"since,omitempty"` // RFC3339 timestamp
	Until     string `json:"until,omitempty"` // RFC3339 timestamp
	DryRun    bool   `json:"dry_run,omitempty"`
}

// Validate validates the replay events request
func (r *ReplayEventsRequest) Validate() error {
	// At least one filter must be specified to prevent replaying all events
	if r.SessionID == "" && r.EventType == "" && r.Since == "" && r.Until == "" {
		return fmt.Errorf("at least one filter (session_id, event_type, since, or until) must be specified")
	}

	return nil
}

// ReplayEventsResponse represents a response to replay events
type ReplayEventsResponse struct {
	Success        bool   `json:"success"`
	EventsFound    int    `json:"events_found"`
	EventsReplayed int    `json:"events_replayed"`
	EventsFailed   int    `json:"events_failed,omitempty"`
	DryRun         bool   `json:"dry_run"`
	Message        string `json:"message"`
	LastError      error  `json:"last_error,omitempty"`
}
